library/core/src/mem/maybe_uninit.rs

```diff
@@ -209,7 +209,8 @@ use crate::slice;
 ///
 /// # Layout
 ///
-/// `MaybeUninit<T>` is guaranteed to have the same size, alignment, and ABI as `T`:
+/// `MaybeUninit<T>` is guaranteed to have the same layout (size, alignment, and field offsets and layouts)
+/// and ABI as `T`:
```

Comment:
The "field offsets and layouts" part is redundant because layouts are compositional. If the size and alignment is the same as T then by necessity the offset of `T` within `MaybeUninit` is zero. And any specific `T` is a `T`, so its field offsets are its field offsets.

This is no different than `#[repr(transparent)] struct Wrapper(MyStruct)`

Reply:
Yeah, I think this is overly complicating things and feels *more* confusing to me than just saying that the size and alignment are the same. MaybeUninit is usually treated as having a single field itself (T) so it's confusing to me to talk about field offset**s** and layout**s**, especially after size/alignment (which is all a layout is).

Reply:
The problem as I see it is twofold:
- While we technically define "layout" to refer to size, alignment, and field offset/layout [1], it's colloquially often used to refer only to size and alignment, and so I think being verbose may help here
- While const generics aren't stable today, you could define `MaybeUninit` as follows, which would be consistent with the size and alignment properties, but would *not* be consistent with the field offset/layout properties:

  ```rust
  union MaybeUninit<T> {
      uninit: (),
      value: [u8; size_of::<T>()],
  }
  ```

[1] I know this was officially documented within the past year, but I can't figure out where we did that...

Reply:
> * While const generics aren't stable today, you could define `MaybeUninit` as follows, which would be consistent with the size and alignment properties, but would _not_ be consistent with the field offset/layout properties:
>   ```rust
>   union MaybeUninit<T> {
>       uninit: (),
>       value: [u8; size_of::<T>()],
>   }
>   ```

This would be a breaking change because the alignment of this type would be the same as u8 and not the same as T. [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=75ebaf90d708b30b40ee3f448eee0eda)

Reply:
That type also differs from the true `MaybeUninit` in terms of preserving all padding bytes (which true `MaybeUninit` does not preserve).